File: src/main/java/org/manapart/item_filters/EndersortEntity.java
package org.manapart.item_filters;

import net.minecraft.inventory.DoubleSidedInventory;
import net.minecraft.inventory.IInventory;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.ChestTileEntity;
import net.minecraft.tileentity.HopperTileEntity;
import net.minecraft.tileentity.TileEntityType;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.SoundCategory;
import net.minecraft.util.SoundEvents;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.text.ITextComponent;
import net.minecraft.util.text.StringTextComponent;

import java.util.HashSet;
import java.util.Set;


public class EndersortEntity extends ChestTileEntity {
    private int transferCooldown = -1;
    private int cooldownBuffer = 5;
    private ChestFinder chestFinder = new ChestFinder();
    private int containerIndex = -1;
    private boolean distributedItems = false;

    public EndersortEntity() {
    }

    @Override
    public TileEntityType<?> getType() {
        return Endersort.tileType;
    }

    @Override
    public void tick() {
        --this.transferCooldown;
        if (this.transferCooldown <= 0 && !this.isEmpty()) {
            this.transferCooldown = cooldownBuffer;
            if (chestFinder.isSearching()) {
                chestFinder.findContainers(this.getLevel(), this.worldPosition);
            } else if (chestFinder.hasContainers()) {
                distributeItems();
            } else {
                chestFinder.resetSearch(worldPosition);
            }
        }
    }

    @Override
    protected ITextComponent getDefaultName() {
        return new StringTextComponent("Endersort");
    }

    public void clearContainers() {
        this.chestFinder.resetSearch(worldPosition);
        this.containerIndex = -1;
    }

    private void distributeItems() {
        containerIndex++;
        if (containerIndex >= chestFinder.getContainerPositions().size()) {
            containerIndex = 0;
        }
//        System.out.println("Distributing Items to " + containerIndex);
        BlockPos pos = chestFinder.getContainerPositions().get(containerIndex);
        IInventory chest = HopperTileEntity.getContainerAt(this.getLevel(), pos);
        if (chest instanceof ChestTileEntity || chest instanceof DoubleSidedInventory) {
            pushItems(chest, pos);
        } else {
            clearContainers();
        }
        if (distributedItems) {
            distributedItems = false;
            getLevel().playSound(null, pos, SoundEvents.ENDERMAN_TELEPORT, SoundCategory.PLAYERS, 1f, 1f);
        }
    }

    private void pushItems(IInventory chest, BlockPos pos) {
        if (!chest.isEmpty()) {
            Set<ResourceLocation> matches = buildMatchMap(chest);
            for (ItemStack item : this.getItems()) {
                attemptToPush(item, chest, matches, pos);
            }
        }
    }

    private Set<ResourceLocation> buildMatchMap(IInventory chest) {
        Set<ResourceLocation> matches = new HashSet<>();
        for (int i = 0; i < chest.getContainerSize(); i++) {
            ItemStack destItem = chest.getItem(i);
            if (!destItem.isEmpty()) {
                matches.add(destItem.getItem().getRegistryName());
            }
        }
        return matches;
    }

    private void attemptToPush(ItemStack item, IInventory destinationInventory, Set<ResourceLocation> matches, BlockPos pos) {
        ResourceLocation matchName = item.getItem().getRegistryName();
        if (!item.isEmpty() && matches.contains(matchName)) {
            //Try to combine stacks first, then try full items
            pushStackable(item, destinationInventory, matchName);
            if (!item.isEmpty()) {
                pushNewStack(item, destinationInventory);
            }

//            if (item.isEmpty()) {
//                System.out.println("Distributed " + matchName + " to " + pos.toShortString());
//            }
        }
    }

    private void pushStackable(ItemStack item, IInventory destinationInventory, ResourceLocation matchName) {
        if (item.isStackable()) {
            for (int i = 0; i < destinationInventory.getContainerSize(); i++) {
                ItemStack destItem = destinationInventory.getItem(i);
                if (destItem.isStackable() && matchName.equals(destItem.getItem().getRegistryName())) {
                    int itemCount = Math.min(destItem.getMaxStackSize() - destItem.getCount(), item.getCount());
                    if (itemCount > 0) {
                        destItem.setCount(destItem.getCount() + itemCount);
                        item.setCount(item.getCount() - itemCount);
                        if (destItem.isEmpty()) {
                            destinationInventory.setChanged();
                        }
                        distributedItems = true;
                        break;
                    }
                }
            }
        }
    }

    private void pushNewStack(ItemStack item, IInventory destinationInventory) {
        for (int i = 0; i < destinationInventory.getContainerSize(); i++) {
            ItemStack destItem = destinationInventory.getItem(i);
            if (destItem.isEmpty()) {
                int itemCount = Math.min(destItem.getMaxStackSize() - destItem.getCount(), item.getCount());
                if (itemCount > 0) {
                    destinationInventory.setItem(i, item.copy());
                    destItem = destinationInventory.getItem(i);
                    destItem.setCount(itemCount);
                    setChanged();
                    item.setCount(item.getCount() - itemCount);
                    if (destItem.isEmpty()) {
                        destinationInventory.setChanged();
                    }
                    distributedItems = true;
                    break;
                }
            }
        }
    }

}

File: src/main/java/org/manapart/endersort/EndersortEntity.kt
package org.manapart.endersort

import net.minecraft.core.BlockPos
import net.minecraft.network.chat.MutableComponent
import net.minecraft.network.chat.TextComponent
import net.minecraft.resources.ResourceLocation
import net.minecraft.sounds.SoundEvents
import net.minecraft.sounds.SoundSource
import net.minecraft.world.CompoundContainer
import net.minecraft.world.Container
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.entity.ChestBlockEntity
import net.minecraft.world.level.block.entity.HopperBlockEntity
import net.minecraft.world.level.block.state.BlockState
import kotlin.math.min

class EndersortEntity(pos: BlockPos, state: BlockState) : ChestBlockEntity(pos, state) {
    private var transferCooldown = -1
    private val cooldownBuffer = 5
    private val chestFinder = ChestFinder()
    private var containerIndex = -1
    private var distributedItems = false
    private val storageHints: MutableMap<ResourceLocation, BlockPos> = mutableMapOf()

    override fun getType(): BlockEntityType<*> = Endersort.tileType
    override fun getName(): MutableComponent = TextComponent("")

    override fun getDefaultName(): TextComponent = TextComponent("Endersort")

    companion object {
        fun sortItems(world: Level?, pos: BlockPos?, state: BlockState?, entity: ChestBlockEntity) {
            (entity as EndersortEntity).tickInternal()
        }
    }

    fun tickInternal() {
        --transferCooldown
        if (transferCooldown <= 0 && !this.isEmpty) {
            transferCooldown = cooldownBuffer
            if (chestFinder.isSearching()) {
                chestFinder.findContainers(getLevel(), worldPosition)
            } else if (chestFinder.hasContainers()) {
                distributeItemsByHints()
                distributeItems()
            } else {
                chestFinder.resetSearch(worldPosition)
            }
        }
    }

    private fun clearContainers() {
        chestFinder.resetSearch(worldPosition)
        storageHints.clear()
        containerIndex = -1
    }

    private fun distributeItemsByHints() {
        if (storageHints.isNotEmpty()) {
            for (item in items) {
                if (!item.isEmpty) {
                    val matchName = item.item.registryName
                    val pos = storageHints[matchName]
                    if (pos != null) {
                        val chest = HopperBlockEntity.getContainerAt(getLevel()!!, pos)
                        if (chest is ChestBlockEntity || chest is CompoundContainer) {
                            val matches = buildMatchMap(chest)
                            attemptToPush(item, chest, matches, pos)
                        } else {
                            println("Found invalid chest at " + pos.toShortString())
                            clearContainers()
                        }
                    }
                }
            }
        }
    }

    private fun distributeItems() {
        if (!chestFinder.hasContainers()) {
            return
        }
        containerIndex++
        if (containerIndex >= chestFinder.containerPositions.size) {
            containerIndex = 0
        }
        //        System.out.println("Distributing Items to " + containerIndex);
        val pos = chestFinder.containerPositions[containerIndex]
        val chest = HopperBlockEntity.getContainerAt(getLevel()!!, pos)
        if (chest is ChestBlockEntity || chest is CompoundContainer) {
            pushItems(chest, pos)
        } else {
            clearContainers()
        }
        if (distributedItems) {
            distributedItems = false
            getLevel()?.playSound(null, pos, SoundEvents.ENDERMAN_TELEPORT, SoundSource.PLAYERS, 1f, 1f)
        }
    }

    private fun pushItems(chest: Container, pos: BlockPos) {
        if (!chest.isEmpty) {
            val matches = buildMatchMap(chest)
            for (item in items) {
                attemptToPush(item, chest, matches, pos)
            }
        }
    }

    private fun buildMatchMap(chest: Container): Set<ResourceLocation?> {
        val matches: MutableSet<ResourceLocation?> = HashSet()
        for (i in 0 until chest.containerSize) {
            val destItem = chest.getItem(i)
            if (!destItem.isEmpty) {
                matches.add(destItem.item.registryName)
            }
        }
        return matches
    }

    private fun attemptToPush(item: ItemStack, destinationInventory: Container, matches: Set<ResourceLocation?>, pos: BlockPos) {
        val matchName: ResourceLocation = item.item.registryName!!
        if (!item.isEmpty && matches.contains(matchName)) {
            //Try to combine stacks first, then try full items
            pushStackable(item, destinationInventory, matchName)
            if (!item.isEmpty) {
                pushNewStack(item, destinationInventory)
            }
            if (item.isEmpty) {
                storageHints[matchName] = pos
                //                System.out.println("Distributed " + matchName + " to " + pos.toShortString());
            }
        }
    }

    private fun pushStackable(item: ItemStack, destinationInventory: Container, matchName: ResourceLocation?) {
        if (item.isStackable) {
            for (i in 0 until destinationInventory.containerSize) {
                val destItem = destinationInventory.getItem(i)
                if (destItem.isStackable && matchName == destItem.item.registryName) {
                    val itemCount = min(destItem.maxStackSize - destItem.count, item.count)
                    if (itemCount > 0) {
                        destItem.count = destItem.count + itemCount
                        item.count = item.count - itemCount
                        if (destItem.isEmpty) {
                            destinationInventory.setChanged()
                        }
                        distributedItems = true
                        break
                    }
                }
            }
        }
    }

    private fun pushNewStack(item: ItemStack, destinationInventory: Container) {
        for (i in 0 until destinationInventory.containerSize) {
            var destItem = destinationInventory.getItem(i)
            if (destItem.isEmpty) {
                val itemCount = min(destItem.maxStackSize - destItem.count, item.count)
                if (itemCount > 0) {
                    destinationInventory.setItem(i, item.copy())
                    destItem = destinationInventory.getItem(i)
                    destItem.count = itemCount
                    setChanged()
                    item.count = item.count - itemCount
                    if (destItem.isEmpty) {
                        destinationInventory.setChanged()
                    }
                    distributedItems = true
                    break
                }
            }
        }
    }

}